### PR TITLE
Add metric white and blacklisting

### PR DESCRIPTION
**Pull request by danrubenstein**
Continued conversation from #146.

I have a specific few questions here about this, as well as looking for general feedback.

1. How should these classes handle regex match errors? 

Looking through the [regex source](https://golang.org/src/regexp/regexp.go), I started reading the docs for `compile` (line 169) before my eyes started to glaze over. Since metrics namespaces are dynamically allocated (e.g., we don't have a natural hook to all of them at the create time of the metric counter). A few options for consideration:

- Compare it against an empty string (or `"benthos"` if we wanted to be clever), and if it threw any errors, we could reject the metric creation entirely.
- Silently reject (in the case of whitelist) or accept (for blacklist) any metrics, and rely on other metrics, and then spit out a log line saying "WARNING this path comparison failed" (or something to that effect).

2. How should these things be tested?

I'm not super familiar with how in Go, we would test that the Whitelist and Blacklist collectors are returning the right struct - I got a little ways into that rabbit hole (ad86dfc), but I got through down in reading about the [Reflect package](https://golang.org/pkg/reflect/#Kind.String) and didn't see anything super helpful - it looks to me that this library is mostly for Go's primitive types, rather than structs.

Anyways, thanks for the time and consideration.

### CODE REVIEW

**Comment by Jeffail** [comment]
Hey @danrubenstein, looking good.

You can compile the regular expression patterns during construction using [`Compile`](https://golang.org/pkg/regexp/#Compile), which would expose broken patterns during service initialization. Then when metrics are registered you would only have to call [`MatchString`](https://golang.org/pkg/regexp/#Regexp.MatchString) which doesn't return errors.

For unit tests it might be easier to create a mock implementation of a metric type that simply records metric paths that are registered to a map.

**Comment by danrubenstein** [comment]
Thanks for the feedback - this was very helpful. 

I went the `compile` route for patterns, so I can store them in the struct at initialization.

I also went the recommended route for unit tests -- creating a somewhat fake `http_server` in the Whitelist and Blacklists. This does mean that the test set up is a little bit extensive, but without this I'm not sure I can get to the underlying `Local` map.

Let me know what you think - thanks!

**Comment by Jeffail** [comment]
Looks good so far, I'll do a proper comb through this tomorrow if that's okay.

I've also implemented the plumbing needed to get nicer metrics documentation and config examples, got it on a branch here: https://github.com/Jeffail/benthos/tree/feature/metrics-docs, you don't need to do anything as I'll merge after this and will fix anything that needs to change as I go.

**Comment by danrubenstein** [comment]
No rush, thanks for taking a look whenever.

The plumbing looks good too - now that there's getting to be a few more metrics options, having the docs and example ready to go is great.

File: lib/metrics/blacklist.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+// SetLogger sets the logger used to print connection errors.
+func (h *Blacklist) SetLogger(log log.Modular) {
+	h.s.SetLogger(log)
+}
+
+// Close stops the Statsd object from aggregating metrics and cleans up
+// resources.
+func (h *Blacklist) Close() error {
+	h.s.Close()
```
Comment (Jeffail) [inline lib/metrics/blacklist.go]: There might be an error returned from the child so change this to `return h.s.Close()`

File: lib/metrics/blacklist.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+// HandlerFunc returns an http.HandlerFunc for accessing metrics for appropriate
+// child types
+func (h *Blacklist) HandlerFunc() http.HandlerFunc {
+	if wHandlerFunc, ok := h.s.(WithHandlerFunc); ok {
+		return wHandlerFunc.HandlerFunc()
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(400)
```
Comment (Jeffail) [inline lib/metrics/blacklist.go]: Return a 501 here instead

File: lib/metrics/blacklist_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2014 Ashley Jeffs
```
Comment (Jeffail) [inline lib/metrics/blacklist_test.go]: Change name and year

**Comment by danrubenstein** [comment]
Thanks for the feedback - I think I caught all of the updates you flagged, but if I missed any, please let me know.

**Comment by Jeffail** [comment]
Hey @danrubenstein, 

This is merged but I used the command tools to squash it in a way that preserves you as the author, but the consequence of that is this pull request doesn't get the satisfying `merged` tag.

Commit: https://github.com/Jeffail/benthos/commit/a1004070b3569d988a07be906e55fafa34191c5a

Docs are being generated now: https://github.com/Jeffail/benthos/tree/master/docs/metrics#blacklist (made some small tweaks to clarify prefix behaviour)

These components are awesome, thank you!

**Comment by danrubenstein** [comment]
Haha no worries - glad we were able to get this feature out there. Thanks for your help!

**Pull request closed without merging**